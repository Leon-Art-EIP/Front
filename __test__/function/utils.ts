import type { ReactTestRendererJSON } from "react-test-renderer";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { expect } from "vitest";

function toJson(component: renderer.ReactTestRenderer): ReactTestRendererJSON {
  const result = component.toJSON();
  expect(result).toBeDefined();

  return result as ReactTestRendererJSON;
}

export function testSnapshot(element: JSX.Element): void {
  const component = renderer.create(element);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
}
