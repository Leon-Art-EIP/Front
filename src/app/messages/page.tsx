export default function Page(): JSX.Element {
  return <div>Page de messages</div>;
}
