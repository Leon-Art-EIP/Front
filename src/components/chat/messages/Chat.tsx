export interface ChatProps {
  content: string;
  sender: 0 | 1;
  dateTime: string;
}

export function Chat(props: ChatProps): JSX.Element {
  /* c8 ignore start */
  return (
    <div className={`flex flex-col ${props.sender ? "items-start" : "items-end"} justify-center px-6`}>
      <div
        className={`rounded-2xl p-4 max-w-xl break-words ${
          props.sender ? "bg-background-hl text-tertiary" : "bg-primary text-white"
        }`}
      >
        {props.content}
      </div>
      {props.dateTime && <span className="text-sm px-2 py-1 text-tertiary">{props.dateTime}</span>}
    </div>
  );
}
/* c8 ignore stop */
