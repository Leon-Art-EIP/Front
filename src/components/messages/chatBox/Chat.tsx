export interface ChatProps {
  content: string;
  sender: 0 | 1; 
  dateTime: string;
}

export function Chat(props: ChatProps): JSX.Element {
  return (
    <div className={`flex flex-col ${props.sender ? "items-start" : "items-end" } justify-center px-6`}>
      <div className={`rounded-2xl p-4 max-w-xl ${props.sender ? "bg-[#E11C0A] text-white" : "bg-[#F3F3F3] text-black" }`}>
        {props.content}
      </div>
      {props.dateTime && <span className="text-md px-2 py-1 text-[#858585]">{props.dateTime}</span>}
    </div>
  );
}
