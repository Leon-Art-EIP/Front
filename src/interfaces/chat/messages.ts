export interface IMessage {
  id: number;
  senderId: string;
  contentType: "text" | "image";
  content: string;
  dateTime: string;
  read: boolean;
}

export interface IOrderInfos {
  userRole: string;
  orderState: string;
  orderRating: number;
  orderPicture: string;
  orderDescription: string;
  orderPrice: number;
}