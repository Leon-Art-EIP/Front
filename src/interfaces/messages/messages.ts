export interface IMessages {
  messages: IMessage[];
}

export interface IMessage {
  id: number;
  sender: number;
  contentType: string;
  content: string;
  date: string;
  read: boolean;
}