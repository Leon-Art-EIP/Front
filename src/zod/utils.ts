export const getValueOrUndefined = (arg: unknown) =>
  arg === "" || arg === undefined || arg === null ? undefined : arg;
