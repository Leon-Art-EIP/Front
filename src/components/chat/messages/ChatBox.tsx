import { useEffect, useRef } from "react";
import { useChat } from "../../../contexts/ChatContext";
import { Chat } from "./Chat";

export function ChatBox(): JSX.Element {
  /* c8 ignore start */
  const { messages, currentUser } = useChat() || {};

  const chatBoxRef = useRef<HTMLDivElement>(null);
  let prevDateTime: Date | null = null;

  useEffect(() => {
    if (chatBoxRef.current) {
      chatBoxRef.current.scrollTop = chatBoxRef.current.scrollHeight;
    }
  }, [messages]);

  const formatDate = (string: string) => {
    const options: Intl.DateTimeFormatOptions = { year: "numeric", month: "long", day: "numeric" };
    return new Date(string).toLocaleDateString([], options);
  };

  const formatTime = (string: string) => {
    return new Date(string).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
  };

  return (
    <div ref={chatBoxRef} dir="btt" className="flex flex-col h-full w-full overflow-y-auto gap-4">
      {messages.map((message, index) => {
        const currentDateTime = new Date(message.dateTime);
        let showFullDate = false;

        if (prevDateTime) {
          if (currentDateTime.toDateString() !== prevDateTime.toDateString()) {
            showFullDate = true;
          }
        } else {
          showFullDate = true;
        }

        prevDateTime = currentDateTime;

        return (
          <div key={index} className="flex flex-col">
            {showFullDate && (
              <>
                <span className="self-center mt-3 h-1 w-1/3 bg-background-hl rounded-full" />
                <span className="text-lg text-center p-2 text-background-hl">{formatDate(message.dateTime)}</span>
              </>
            )}
            <Chat
              dateTime={formatTime(message.dateTime)}
              key={message.id}
              content={message.content}
              sender={message.senderId === currentUser?.user.id ? 0 : 1}
            />
          </div>
        );
      })}
    </div>
  );
}
/* c8 ignore stop */
