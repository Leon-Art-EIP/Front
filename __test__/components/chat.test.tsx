import { test } from "vitest";
import { testSnapshot } from "../function/utils";
import { Chat } from "../../src/components/chat/messages/Chat";

test("Chat sender 0 snapshot", () => {
  testSnapshot(<Chat content="content" dateTime="dateTime" sender={0} />);
});

test("Chat sender 1 snapshot", () => {
  testSnapshot(<Chat content="content" dateTime="dateTime" sender={1} />);
});
