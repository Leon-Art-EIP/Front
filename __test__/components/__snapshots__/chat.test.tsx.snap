// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chat sender 0 snapshot 1`] = `
<div
  className="flex flex-row justify-end items-center px-6"
>
  <div
    className="rounded-2xl p-4 max-w-xl break-words bg-primary text-white"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <p
      className="mb-1"
    >
      content
    </p>
  </div>
</div>
`;

exports[`Chat sender 1 snapshot 1`] = `
<div
  className="flex flex-row justify-start items-center px-6"
>
  <div
    className="rounded-2xl p-4 max-w-xl break-words bg-background-hl text-tertiary"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <p
      className="mb-1"
    >
      content
    </p>
  </div>
</div>
`;
